File: client-android/serenada-core/src/test/java/app/serenada/core/fakes/FakeSignaling.kt
package app.serenada.core.fakes

import app.serenada.core.call.SessionSignaling
import app.serenada.core.call.SignalingMessage

class FakeSignaling : SessionSignaling {
    override var listener: SessionSignaling.Listener? = null

    val connectCalls = mutableListOf<String>()
    val sentMessages = mutableListOf<SignalingMessage>()
    var closeCalls = 0
        private set
    var connected = false
        private set

    override fun connect(host: String) {
        connectCalls.add(host)
    }

    override fun isConnected(): Boolean = connected

    override fun send(message: SignalingMessage) {
        sentMessages.add(message)
    }

    override fun close() {
        closeCalls++
        connected = false
    }

    override fun recordPong() {}

    // ── Test drivers ──

    fun simulateOpen(transport: String = "ws") {
        connected = true
        listener?.onOpen(transport)
    }

    fun simulateMessage(message: SignalingMessage) {
        listener?.onMessage(message)
    }

    fun simulateClosed(reason: String = "test") {
        connected = false
        listener?.onClosed(reason)
    }

    fun sentMessages(ofType: String): List<SignalingMessage> =
        sentMessages.filter { it.type == ofType }
}

File: client-ios/SerenadaCore/Tests/SerenadaCoreTests/Fakes/FakeSignaling.swift
import Foundation
@testable import SerenadaCore

@MainActor
final class FakeSignaling: SessionSignaling {
    weak var listener: SignalingClientListener?

    private(set) var connectCalls: [String] = []
    private(set) var sentMessages: [SignalingMessage] = []
    private(set) var closeCalls = 0
    private(set) var connected = false

    func connect(host: String) {
        connectCalls.append(host)
    }

    func isConnected() -> Bool { connected }

    func send(_ message: SignalingMessage) {
        sentMessages.append(message)
    }

    func close() {
        closeCalls += 1
        connected = false
    }

    func recordPong() {}

    // MARK: - Test Drivers

    func simulateOpen(transport: String = "ws") {
        connected = true
        listener?.onOpen(activeTransport: transport)
    }

    func simulateMessage(_ message: SignalingMessage) {
        listener?.onMessage(message)
    }

    func simulateClosed(reason: String = "test") {
        connected = false
        listener?.onClosed(reason: reason)
    }

    func sentMessages(ofType type: String) -> [SignalingMessage] {
        sentMessages.filter { $0.type == type }
    }
}
